wordpress-mobile/WordPress-Android
# Fix 'ghost sites' on logout

Fixes #5838 (FluxC issue: https://github.com/wordpress-mobile/WordPress-FluxC-Android/issues/433). The required FluxC changes have already been reviewed (https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/438) and merged to a feature branch, `wpcom-login-logout-changes`, which should be merged into `develop` (and this PR updated to use the latest FluxC hash) before merging this PR.

This PR isn't needed to fix the issue exactly - the fix happened in FluxC, but as a result it now requires an account to be present in the store before WP.com sites can be stored. This PR updates sign in, account creation and migration to wait for the account to be fetched before fetching sites, to ensure that the sites aren't rejected by FluxC due to the lack of account.

**To test:**
First reproduce the original issue as it's a little tricky:
1. Sign in to both a WordPress.com account and a self-hosted site
2. Change [this](https://github.com/wordpress-mobile/WordPress-Android/blob/ae3f9c922e4e03f43a4324a921d875eccd2fe861/WordPress/src/main/java/org/wordpress/android/WordPress.java#L117) to 1 second
3. Visit the Me screen, and background the app
4. Open the app and immediately press 'Log out of WordPress.com'
5. Re-opening the app should trigger a FETCH_SITES call, and the goal is to trigger a sign out before that network call completes (a slower network helps to pull this off, and you can verify the order of actions in the logs)
6. Open the site list - you should still have all your WordPress.com sites, even though you just signed out
7. Selecting any WordPress.com site seems to work, but all attempts to fetch data fail

Doing the same with this patch should remove all the sites correctly. You should see `Can't insert WP.com site ..., missing user account` messages in the logs if you correctly trigger a sign out before `FETCH_SITES` completes.

Additionally, please test:
1. Signing in to WP.com and self-hosted sites
2. New account creation
3. Migration (from 6.9 or earlier)

cc @maxme and heads up @hypest
